library/std/src/sys/unix/locks/futex.rs

```diff
@@ -162,3 +164,87 @@ impl Condvar {
         r
     }
 }
+
+/// A reentrant mutex. Used by stdout().lock() and friends.
+///
+/// The 'owner' field tracks which thread has locked the mutex.
+///
+/// We use current_thread_unique_ptr() as the thread identifier,
+/// which is just the address of a thread local variable.
```

Comment:
That might not work if a thread doesn't unlock and then exits. Another new thread may then suddenly acquire the lock if the allocator happens to return the same pointer.

Reply:
> That might not work if a thread doesn't unlock and then exits. Another new thread may then suddenly acquire the lock if the allocator happens to return the same pointer.

That should work out fine, as the old thread doesn't exist anymore, it's fine for the new thread to take over ownership of the lock.

> There is a ThreadId type you can use instead.

I am aware, but the ThreadId type is 64-bit, which is problematic on some 32-bit platforms.